### ISSUE TITLE
lab3 review

### ISSUE THREAD

**Issue by shadow036** [vmask25/Computational-Intelligence-CI-2022#8]
Hello, I'm going to leave a review explaining my thoughts on your implementation of the third lab.

# Evolutionary strategy
- The "pick_even_max" and the "pick_odd_max" select respectively the group of rows having even or odd number of elements and then from these groups the row having the **highest index** is chosen (as in the row closer to the bottom). I don't know if that was intended or not but based on the comments I expected to see a strategy that was choosing the row with even/odd elements and the highest number of objects left among all the rows having even/odd number of objects left.
- In line 91-92 you can spare 1 line by using (maybe in this case a tuple would have been conceptually more correct):

        champion = [result, individual]
- Following this thought process you discarded the concept of population and decided to go for a series on rounds in which a single individual is evaluated before replacing it with the next one. As a consequence also the concept of crossover was different as each new indivudual is not the product of other parents individuals. Anyway the final result would have been the same (since at the beginning of each generation you still generate a new individual with "good" genes in average) so I guess that's ok. 
- The "champion" variable In line 78 is redundant
- We can discuss if the value of 0.3 as RESULT_TRESH is ok in this scenario (it may correspond to a low winrate threshold against the dumb strategy and a very high one against the optimal strategy).
- The "evolve_genome" function behaves in an asymmetric way depending on the comparison between the "results"  variable and the RESULT_TRESH  constant: I don't know if this was intended and if it was it would have been better to explain the reasoning for that in the README.
- Instead of using a threshold (RESULT_TRESH) maybe you could have based the decision of improving the genes probabilities on the winrate difference with respect to the previous individual. On the other hand in this way you could have found yourself stuck in a local minimum.
- In order to shrink the code you could have replaced the "evaluate" function with:
    
        def evaluate(player1, player2):
            won = 0
            for m in range(NUM_MATCHES):
                board = Nim(NIM_SIZE)
                player = 0
                i = random.randint(0,1)
                while board:
                    flag = i % 2 != 0
                    opponent = (player1 if flag else player2).move()
                    player = int(not flag)
                    ply = opponent(board)
                    board.nimming(ply)
                    i+=1
                won += int(not bool(player))
            return won/NUM_MATCHES
Overall I think that's a good implementation of an evolutionary algorithm even though the generation of a new individual ("generate_individual" function) is bit clunky.
# Minmax strategy
- The "evaluation" function can be shrunk in this way, since a boolean value is converted to an integer when it is involved in a arithmetic operation:

        def evaluation(state):
            return state - 1
- Also in line 54 you could have just used "val" as a boolean since all values different from 0 corresponds to "True":
    
        if val:
            return None, val
- Even though the mechanism you used to find a winning "path" is very clever (I'm referring to the +1 value during even turns and -1 during odd ones), it was very hard for me to understand at first and for this reason maybe it would have been better to add a more detailed description about the procedure in the README. 

In conclusion, I really like how you managed to solve this problem with a relatively small code and that you also managed to include two very clever optimization techniques.
Even thogh it's not a complete min-max due to the presence of random moves, it is certainly a very good tradeoff in order to avoid huge memory consumption.

# Reinforcement learning strategy
- In line 69 (in agent.py)  the agent is already aware of all possible actions due to the previous "init_reward" function and so there should be no "if ply in self.G", together with the body in the "else" block.

Even though it is slower I think that setting the rows as state instead of plies makes more sense but of course you still needed to find a tradeoff with the computational speed.
